### DISCUSSION TITLE
Performing a fast and precise swipe/scroll

### DISCUSSION THREAD

**Discussion by dragojs** [appium/appium#17468] [Q&A]
Hi,

Does anyone know how to build a precise method for doing a fast and **precise** swipe/scroll ?

By precise, I mean swiping/scrolling from A to B as fast as possible without inertia to keep moving my screen after the swipe/scroll is done executing.

The closest I got was via something like

`
Driver.ExecuteScript(
                "mobile: dragGesture",
                new Dictionary<string, int>
                {
                    { "startX", startx },
                    { "startY", starty },
                    { "endX", endx },
                    { "endY", endy },
                    { "speed", speed }
                });
`

where speed is = Math.Max(window.Width, window.Height) / 4;

**Comment by jlipps** [comment]
I would suggest just doing what a user would do. How would a user keep the screen from continuing to scroll? You'd need to construct an action with a sort of fishhook shape and timed appropriately. You can express something like this using the Actions API.

**Reply by dragojs** [reply]
thought about that myself, seems to work in a preliminary phase. I'll come back if I find a satisfactory implementation

**Reply by dragojs** [reply]
Trying to use something like this:
new TouchAction(AppiumManager.Driver).Press(startx, starty).MoveTo(endx, endy).Wait(100).Release().Perform();

But not getting anywhere. It swipes perfectly a few times with specific points (don't even need a fish hook shape), then doesn't want to swipe at all with a different set of points.
Either it doesn't want to start the swipe when the start points are on top of an element or I don't know what the problem is. Not seeing anything out of the ordinary in appiums logs.

**Comment by dragojs** [comment]
`new TouchAction(AppiumManager.Driver)
                .Press(startx, starty)
                .MoveTo(fishhookx, fishhooky)
                .MoveTo(endx, endy)
                .Release().Perform();`
I've found this to work.
The first MoveTo transforms my swipe like this:
![image](https://user-images.githubusercontent.com/48386392/189948972-e72e17e0-3916-4423-819d-10a8dc4b0fa7.png)
The problem I've got with this is, the _fishhook_ makes my swipe go +50 units on the x and y axes towards **a'** before going in the direction that I want. There's a certain size threshold for this miniswipe towards a'; too few units and my desired swipe won't work at all; too many units and it looks clunky.

Either way, I;m not very keen to use this solution as I don't understand why this won't work without the little fishhook swipe. I suspect a timing issue of some sort but I don't have time to investigate.

**Reply by jlipps** [reply]
I'm not sure whether you're using the right client methods. It looks like you're using .net? Are these the w3c actions or some other thing? I also don't see any duration information for any of these, which is another big item that can be tweaked for stability.

**Reply by dragojs** [reply]
I am indeed using .net with these methods: https://appium.io/docs/en/writing-running-appium/touch-actions/ .
It is unclear to me whether these are w3c actions. Appium seems to be doing this:
![image](https://user-images.githubusercontent.com/48386392/190332554-a9e90978-405d-45ad-8b20-8b0d39b9c69e.png)

Indeed, I have not added any .Wait()s as I have not seen it cause any different behaviours.

Upon further reading I discovered I may have to use http://appium.io/docs/en/commands/interactions/actions/ as TouchActions will be deprecated. I'll try and see if this is any different..

**Reply by RSM-SBA** [reply]
As the touchActins are marked as deprecated (and not supported anymore in selenum 4) you should better use the actions API.
https://www.selenium.dev/documentation/webdriver/actions_api/

*Oh missed your last comment which mentioned the same

**Reply by dragojs** [reply]
Following the example from here:
http://appium.io/docs/en/commands/interactions/actions/
I created the following action list of swipes:
`
var inputDevice = new OpenQA.Selenium.Appium.Interactions.PointerInputDevice(PointerKind.Touch);
var actionSequence = new ActionSequence(inputDevice, 0);

actionSequence.AddAction(inputDevice.CreatePointerMove(CoordinateOrigin.Viewport, startx, starty, TimeSpan.FromSeconds(1)));
actionSequence.AddAction(inputDevice.CreatePointerDown(PointerButton.TouchContact));
// actionSequence.AddAction(inputDevice.CreatePause(TimeSpan.FromMilliseconds(1000)));
actionSequence.AddAction(inputDevice.CreatePointerMove(CoordinateOrigin.Viewport, endx, endy, TimeSpan.FromSeconds(1)));
actionSequence.AddAction(inputDevice.CreatePause(TimeSpan.FromSeconds(1)));
actionSequence.AddAction(inputDevice.CreatePointerUp(PointerButton.TouchContact));

AppiumManager.Driver.PerformActions(new List<ActionSequence> { actionSequence });
`

By fiddling a bit with it, I discovered that the pauses and the duration inside the CreatePointer methods are important and cannot be arbitrary for the swipe to work properly.
I also discovered by turning on the Pointer Location setting from the device's Developer Options that this swipe looks to be going well for ~90% of it's lifetime, gets stuck (probably a CreatePause kicks in, but why here !?) then continues as if a finger flick happened from that point.

[appium_logs.md](https://github.com/appium/appium/files/9575595/appium_logs.md)
I've attached the logs from appium.

What I can't wrap my head around is what exactly is the right proportion for these in order to get a proper swipe ??


A later edit--

Discovered the following:
- CreatePause brings absolutely no benefit to my use case.
- The faster I make the swipe (via a smaller duration for CreatePointerMove), the more imprecise it is. This seems to appear because the screen doesn't start dragging along with the swipe command instantly; instead it lags behind a little bit, as if I were first pulling a rope attached to the object that must first stretch fully before the object itself is pulled.

I'm trying to settle on the variant below:
Split the swipe in 2; once up to **prefinish** and the second to my desired end. This way I avoid the swipe moving past my end point due to inertia.

`            var inputDevice = new OpenQA.Selenium.Appium.Interactions.PointerInputDevice(PointerKind.Touch);
            var actionSequence = new ActionSequence(inputDevice, 0);

            actionSequence.AddAction(inputDevice.CreatePointerMove(CoordinateOrigin.Viewport, startx, starty, TimeSpan.FromMilliseconds(0)));
            actionSequence.AddAction(inputDevice.CreatePointerDown(PointerButton.TouchContact));
            actionSequence.AddAction(inputDevice.CreatePointerMove(CoordinateOrigin.Viewport, prefinishX, prefinishY, TimeSpan.FromMilliseconds(250)));
            actionSequence.AddAction(inputDevice.CreatePointerMove(CoordinateOrigin.Viewport, endx, endy, TimeSpan.FromMilliseconds(250)));
            actionSequence.AddAction(inputDevice.CreatePointerUp(PointerButton.TouchContact));

            AppiumManager.Driver.PerformActions(new List<ActionSequence> { actionSequence });`

Any explanations for my issue or thoughts are welcome

Even later edit--

The laggy swiping as if pulling a rope thing can even be done manually. Looks like android thinks I'm not actually doing a swipe unless I keep moving in a swipe motion for a few (don't know how many) milliseconds.
When appium does the swipe, and it does it fast, it covers more terrain in that small bubble of milliseconds, hence my swipe is imprecise.

**Reply by jlipps** [reply]
yeah, i would not recommend trying to do very fast swipes. slow and sure is much better for automation

**Reply by vnuta9** [reply]
@dragojs @jlipps 
Dear Gurus;
**I have learnt so much from this post.** This post also resonates very much what I have been experiencing in my test automation on an Android device.
Unfortunately, I am hesitating to take a plunge of installing **appium-gestures-plugin** ver2.0, because it irrevocably puts me in a spot where I have to continue using the new W3C Action classes, about which I don't have much idea, and not much public available example code while I don't have enough time allocated on it to work. So, I need to keep using the deprecated **TouchAction** classes for some more time even though I am on **Appium 2.0.0.beta66**.

Any more wise suggestions you have up your sleeves ?
Thanks for a great 'discussion'.

**Reply by saikrishna321** [reply]
@vnuta9 `I need to keep using the deprecated TouchAction classes` I would not recommend, we will remove these deprecated methods in future release.  We have these deprecated functions for more than a year. So the window to remove these would have soon. 

`and not much public available example code` Not true. We have enough material. Please check these
https://www.youtube.com/watch?v=5-soIXmWMZM&pp=ygUPYXBwaXVtIGdlc3R1cmVz
https://www.headspin.io/blog/automating-complex-gestures-with-the-w3c-actions-api
https://github.com/saikrishna321/VodQaAdvancedAppium/blob/master/src/test/java/com/appium/gesture/GestureTest.java

Way going forward is using W3C Actions API for gestures.

**Reply by vnuta9** [reply]
@saikrishna321 @jlipps 
Appreciate your kind reply and help Sai !
It is much help for me.
I will spend time to assimilate the content of the links provided which are reasonably comprehensive and implement the principles with 'gestures' plugin and develop my gestures utility methods.
BTW, in Sakthivel's talk, I came to know about another worthy plugin called 'Wait' which when ran on the server, would take care of the wait mechanism completely on the server side. Awesome!!

**Reply by vnuta9** [reply]
@saikrishna321 
BTW, which version of Java-client was used in building the 
https://github.com/saikrishna321/VodQaAdvancedAppium/blob/master/src/test/java/com/appium/gesture/GestureTest.java ?
Because I could see the usage of MobileElement and other **non existing classes in the latest java-client ver8.5.**
Appreciate your reply.

**Reply by saikrishna321** [reply]
It has not been updated to the latest java-client. But usage of Actions API is same in the latest version as well

**Reply by vnuta9** [reply]
@saikrishna321 
It is very difficult to contribute these side-educational projects, while contributing to the appium project, plus delivering the regular job duties. It is very selfish to ask like this.
If you intend to update the [VodQaAdvancedAppium](https://github.com/saikrishna321/VodQaAdvancedAppium) project with the latest Javaa-Client library, at what version would it be ? Probably at the introduction of the next round of major changes in the API ???

**Reply by saikrishna321** [reply]
8.5.1 would be the latest update

**Reply by vnuta9** [reply]
@saikrishna321 
Thanks Sai !
While making the changes, it will be much better(for us) if the implementation methods are as much stand alone/independent/generic as possible. Usually you would write a @Test and call the implementation method, as I saw it from the previous code.
I will look forward to the new push to the github master.
